File: DynamicPrograming/ts/quickestConcat.ts
export function quickestConcat(s: string, words: string[]): number {
  const memo = new Map<number, number>();

  const dfs = (i: number): number => {
    if (i === s.length) return 0;
    if (memo.has(i)) return memo.get(i)!;

    let best = Infinity;
    for (const word of words) {
      if (s.startsWith(word, i)) {
        const attempt = 1 + dfs(i + word.length);
        best = Math.min(best, attempt);
      }
    }

    memo.set(i, best);
    return best;
  };

  const result = dfs(0);
  return Number.isFinite(result) ? result : -1;
}

File: DynamicPrograming/python/quickestConcat.py
from functools import lru_cache
from typing import List


def quickestConcat(s: str, words: List[str]) -> int:
    @lru_cache(maxsize=None)
    def dfs(i: int) -> int:
        if i == len(s):
            return 0
        best = float('inf')
        for word in words:
            if s.startswith(word, i):
                best = min(best, 1 + dfs(i + len(word)))
        return best

    result = dfs(0)
    return -1 if result == float('inf') else result
